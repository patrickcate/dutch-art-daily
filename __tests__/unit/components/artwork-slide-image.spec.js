import mockData from '@fixtures/mock-data.js'
import ArtworkSlideImage from '@components/artwork-slide-image.vue'

describe('ArtworkSlideImage Component', () => {
  let wrapper
  let options
  let state

  beforeEach(() => {
    options = {
      propsData: {
        id: '01-01',
      },
    }

    state = {
      state: {
        currentArtworkHeight: 100,
      },
    }

    wrapper = createWrapper(ArtworkSlideImage, options, state)
  })

  it('should be a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBe(true)
  })

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('image load event fires', async () => {
    const imageHasLoadedMock = jest.fn()

    wrapper = createWrapper(ArtworkSlideImage, {
      ...options,
      methods: {
        imageHasLoaded() {
          imageHasLoadedMock()
        },
      },
    })

    wrapper.find('img').trigger('load')
    await wrapper.vm.$nextTick()

    expect(imageHasLoadedMock).toHaveBeenCalled()
  })

  it('slide is tracked and height set after image loads', async () => {
    wrapper = createWrapper(
      ArtworkSlideImage,
      {
        ...options,
      },
      {
        state: {
          currentArtworkHeight: 400,
        },
      }
    )

    wrapper.vm.$el = {
      parentNode: {
        scrollHeight: 1000,
      },
    }

    expect(Object.keys(wrapper.vm.$store.state.loadedSlides).length).toBe(0)
    expect(wrapper.vm.$store.state.currentArtworkHeight).toBe(400)

    wrapper.find('img').trigger('load')
    await wrapper.vm.$nextTick()

    expect(Object.keys(wrapper.vm.$store.state.loadedSlides)[0]).toBe('01-01')
    expect(wrapper.vm.$store.state.currentArtworkHeight).toBe(1000)
  })

  it('only update height from active slide image', async () => {
    wrapper = createWrapper(
      ArtworkSlideImage,
      {
        ...options,
        propsData: {
          id: '12-12',
        },
      },
      {
        state: {
          currentArtworkHeight: 0,
        },
        getters: {
          getArtworkById: () => () => {
            return mockData.artwork
          },
        },
      }
    )

    wrapper.find('img').trigger('load')
    await wrapper.vm.$nextTick()

    // Active id and image prod id don't match.
    expect(wrapper.props('id')).toBe('12-12')
    expect(wrapper.vm.$store.state.activeId).toBe('01-01')

    wrapper.setProps({ id: '01-01' })

    wrapper.vm.$el = {
      parentNode: {
        scrollHeight: 0,
      },
    }

    wrapper.find('img').trigger('load')
    await wrapper.vm.$nextTick()

    expect(wrapper.vm.$el.parentNode).toBeNull()
  })

  it('only generates available image sizes', () => {
    wrapper = createWrapper(ArtworkSlideImage, options, {
      ...state,
      getters: {
        getArtworkById: () => () => {
          return {
            ...mockData.artwork,
            ...{
              xl3: null,
            },
          }
        },
      },
    })

    const dataSrcSet = wrapper.find('img').attributes('data-srcset')
    expect(dataSrcSet).not.toContain('xl3')
  })
})
