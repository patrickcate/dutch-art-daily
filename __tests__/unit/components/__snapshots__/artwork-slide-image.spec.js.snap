// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtworkSlideImage Component renders correctly 1`] = `<div class="artwork-image-frame"><img data-src="/photos/01-01/01-01--xs3-d0wAmzwM620.jpg" data-srcset="/photos/01-01/01-01--xs3-d0wAmzwM620.jpg 640w, /photos/01-01/01-01--xs2-d0wAmzwM620.jpg 800w, /photos/01-01/01-01--xs-d0wAmzwM620.jpg 960w, /photos/01-01/01-01--sm-d0wAmzwM620.jpg 1120w, /photos/01-01/01-01--md-d0wAmzwM620.jpg 1280w, /photos/01-01/01-01--lg-d0wAmzwM620.jpg 1440w, /photos/01-01/01-01--xl-d0wAmzwM620.jpg 1600w, /photos/01-01/01-01--xl2-d0wAmzwM620.jpg 1760w, /photos/01-01/01-01--xl3-d0wAmzwM620.jpg 1920w" data-sizes="(min-height: 700px) calc((4000 / 4485) * (100vh - 280px)), calc((4000 / 4485) * 100vh), 100vw" alt="Johannes Vermeer's painting: The Milkmaid" class="artwork-image swiper-lazy"></div>`;
